### PR TITLE
Popup Widget (Issue #516)

### PR DESCRIPTION
**Pull request by dtsnell4**
Signed-off-by: <Dan.Snell@lexisnexis.com>

### CODE REVIEW

**Comment by dtsnell4** [comment]
@GordonSmith @jbrundage @mlzummo I had to close and reissue this pull request due to merge conflicts with master.

**Comment by dtsnell4** [comment]
@mlzummo I fixed all the issues you commented on

File: src/layout/Popup.js [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        widgets
+            .each(function (d) {
+                var widgetSize = context.widgetSize(d3.select(this));
+                d.resize({ width: widgetSize.width, height: widgetSize.height });
+            })
+        ;
+        widgets.exit().each(function (d) {
+            d.target(null);
+        }).remove();
+
```
Comment (mlzummo) [inline src/layout/Popup.js]: extra sapce

File: src/layout/Popup.js [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    Popup.prototype.publish("bottom", null, "number", "Bottom position property of popup",null,{});
+    Popup.prototype.publish("left", null, "number", "Left position property of popup",null,{});
+    Popup.prototype.publish("right", null, "number", "Right position property of popup",null,{});
+    Popup.prototype.publish("position", "relative", "set", "Value of the 'position' property",["absolute", "relative", "fixed", "static", "initial", "inherit" ],{tags:['Private']});
+    Popup.prototype.publish("showTestButton", false, "boolean", "Test Button in target",null,{});
+    Popup.prototype.publish("buttonName", null, "string", "Button name property",null,{});
+    Popup.prototype.publish("buttonLabel", "", "string", "Button label text",null,{});
+    Popup.prototype.publish("widget", null, "widget", "Widget",null,{tags:['Private']});
+
+
```
Comment (mlzummo) [inline src/layout/Popup.js]: extra space

File: src/layout/Popup.js [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        define(["d3", "../common/HTMLWidget", "../layout/Surface", "../form/Input", "../c3chart/Column", "css!./Popup", "css!font-awesome",], factory);
+    } else {
+        root.layout_Popup = factory(root.d3, root.common_HTMLWidget, root.layout_Surface, root.form_Input, root.c3chart_Column);
+    }
+}(this, function (d3, HTMLWidget, Surface, Input, Column) {
+    function Popup() {
+        HTMLWidget.call(this);
+
+        this._tag = "div";
+
```
Comment (mlzummo) [inline src/layout/Popup.js]: can prolly remove extra spaces here ...

**Comment by dtsnell4** [comment]
@GordonSmith this has been updated with the latest fixes.  Let me know what else you want me to do with it.

**Comment by dtsnell4** [comment]
Fixed merge conflict with master

**Comment by jbrundage** [comment]
@GordonSmith could you pull this into master?

**Comment by jbrundage** [comment]
@dtsnell4 Actually, the latest changes to HTMLWidget.js in master prevent the Popup from setting its 'position', 'left' and 'top' with publish parameters.
https://github.com/hpcc-systems/Visualization/commit/92da9bfe36a488447511f6998addc2c060e85773#diff-686a6ea71d2593f374a3c72ae626bf67R146

**Comment by dtsnell4** [comment]
Yes, I had the same issue with the table widget, will update the PR to fix it.

**Comment by dtsnell4** [comment]
@GordonSmith @jbrundage @mlzummo  Please re-review.

Moved element size/position code to postUpdate()

**Comment by GordonSmith** [comment]
@jbrundage If your happy with this I will pull (since you have been using it?)

**Comment by jbrundage** [comment]
@GordonSmith I'm happy with it

**Pull request merged**